core/shared/src/main/scala/scalaz/zio/Queue.scala

```diff
@@ -3,156 +3,406 @@
+                      .enqueue(as.takeRight(c).toList),
+                    putters,
+                    hook
+                  )
+                }
+              )
+            case Dropping =>
+              (
+                p.complete(false),
+                Surplus(values.enqueue(addToQueue.toList), putters, hook)
+              )
+          }
+        }
+
+      case (p, state @ Shutdown) => (p.interrupt, state)
+    }
+
+    val release: (Boolean, Promise[Nothing, Boolean]) => IO[Nothing, Unit] = {
+      case (_, p) => p.poll.void <> removePutter(p)
+    }
+
+    Promise.bracket[Nothing, State[A], Boolean, Boolean](ref)(acquire)(release)
   }
+
+  /**
+   * Adds a shutdown hook that will be executed when `shutdown` is called.
+   * If the queue is already shutdown, the hook will be executed immediately.
+   */
+  final def onShutdown(io: IO[Nothing, Unit]): IO[Nothing, Unit] =
+    IO.flatten(ref.modify {
+      case Deficit(takers, hook)         => IO.unit -> Deficit(takers, hook *> io)
+      case Surplus(queue, putters, hook) => IO.unit -> Surplus(queue, putters, hook *> io)
+      case state @ Shutdown              => io      -> state
+    })
 }
-object Queue extends Serializable {
+
+object Queue {
 
   /**
    * Makes a new bounded queue.
    * When the capacity of the queue is reached, any additional calls to `offer` will be suspended
    * until there is more room in the queue.
    */
   final def bounded[A](capacity: Int): IO[Nothing, Queue[A]] =
-    Ref[State[A]](Surplus[A](IQueue.empty, IQueue.empty)).map(new Queue[A](capacity, _))
+    createQueue(Some(capacity), BackPressure)
+
+  /**
+   * Makes a new bounded queue with sliding strategy.
+   * When the capacity of the queue is reached, new elements will be added and the old elements
+   * will be dropped.
+   */
+  final def sliding[A](capacity: Int): IO[Nothing, Queue[A]] = createQueue(Some(capacity), Sliding)
+
+  /**
+   * Makes a new bounded queue with the dropping strategy.
+   * When the capacity of the queue is reached, new elements will be dropped.
+   */
+  final def dropping[A](capacity: Int): IO[Nothing, Queue[A]] =
+    createQueue(Some(capacity), Dropping)
 
   /**
    * Makes a new unbounded queue.
    */
-  final def unbounded[A]: IO[Nothing, Queue[A]] = bounded(Int.MaxValue)
+  final def unbounded[A]: IO[Nothing, Queue[A]] = createQueue(None, BackPressure)
+
+  private def createQueue[A](
+    capacity: Option[Int],
+    strategy: SurplusStrategy
+  ): IO[Nothing, Queue[A]] =
+    Ref[State[A]](Surplus[A](IQueue.empty, IQueue.empty, IO.unit)).map(state => new Queue[A](capacity, state, strategy))
+
+  private[zio] object internal {
+
+    sealed trait SurplusStrategy
+
+    case object Sliding extends SurplusStrategy
+
+    case object Dropping extends SurplusStrategy
+
+    case object BackPressure extends SurplusStrategy
+
+    sealed trait State[+A] {
+      def size: IO[Nothing, Int]
+      def shutdownHook: IO[Nothing, Unit]
```

Comment:
The `Shutdown` doesn't need a shutdown hook, so you could remove this from `State` if you wanted and just manually extract it out of `Surplus` or `Deficit`.